### ISSUE TITLE
[Question] Sampling and optimising GPs asynchronously

### ISSUE THREAD

**Issue by cisprague** [cornellius-gp/gpytorch#1145]
Hi,

I need to continuously optimise an approximate GP (in process 1) and sample it and add to the dataset it is regressing (in process 2).

Is there a way to sample a GP while it is being optimised?

**Comment by KeAWang** [comment]
You could just keep switching between eval mode and training mode as you're optimizing the GP model. It would be slow though because you would be recomputing the caches at every iteration

**Comment by joseduc10** [comment]
@KeAWang is this the only way? If we have something like a "flipped" DKL, where data goes through a GP and then the output of the GP is used as input to another model, do we have to put the GP in eval model in every training iteration and switch back to training mode? Do the gradients still back propagate to the GP in this scenario?

**Comment by jacobrgardner** [comment]
If you want to train a GP in eval mode including things like hyperparameters, all you should need to do is do the training in a `with gpytorch.settings.detach_test_caches(False)` context, which will allow training through them.

You used to have to switch between train/eval mode each iteration to clear the caches, but as of version 1.0.0 we clear the caches automatically any time you backpropagate through them (see https://github.com/cornellius-gp/gpytorch/pull/916).

**Comment by joseduc10** [comment]
@jacobrgardner Thank you for the quick reply. I think my setting is slightly different.

I am trying to train a model similar to the one in this paper: https://arxiv.org/pdf/1706.04152.pdf , where the authors use the output of a GP as input to an RNN for classification (but it could be a simpler model, such as a logistic regression). 

So, currently, my training loop (switching between train() and eval()) looks as follows:

```python
niter = 1000
for i in range(niter):
    # Hadamard Multitask GP from 
    #https://docs.gpytorch.ai/en/latest/examples/03_Multitask_Exact_GPs/Hadamard_Multitask_GP_Regression.html
    model_gp.train()
    optimizer.zero_grad()
    outputs = model(full_train_x, full_train_i)
    loss_gp = -mll(output, y))
    # now sample from this model at regular intervals
    model.eval()
    output = model(regular_x, regular_i) 
    # run linear regression
    #TODO input to this model should actually be a set of samples from the GP
    y_pred = model_lr(outputs.mean)
    loss_lr = mse_loss(y_pred, y_true)
    
    loss = loss_gp + loss_lr     # I probably need some weighting parameters here
    loss.backward()
    print('Iter %d/%d - GP Loss: %.3f - MSE Loss: %.3f - Lengthscale: %.3f' 
          % (i + 1, niter, loss_gp.item(), loss_lr.item(), model.data_covar_module.lengthscale.item()))
    optimizer.step()
```

Should I be setting `gpytorch.settings.detach_test_caches(False)` instead of switching between train and eval? Thank you for your help.

**Comment by jacobrgardner** [comment]
No, in your specific application what you've done looks correct as long as you want to be training the GP hypers through a GP MLL and not the final MSE loss. If you want the GP hypers to also get gradients from the MSE loss, then you would add that context.

**Comment by joseduc10** [comment]
OK. I see. In my application, the input to the second model would be samples from the GP using `model.rsample()`, and the loss will be averaged over these samples. So, in that case, I should add the context, but besides that, the training loop stays the same. Is that correct?

By the way, thank you to the GPytorch team for developing this package! I've been using it extensively in the past year. It really makes complex GP models more accessible.

**Comment by XiaoranHao** [comment]
@jacobrgardner Hi! I am calculating the gradients of hypers from the MSE loss with gpytorch.settings.detach_test_caches(False). But I found that I still need to switch between train mode and eval mode. Otherwise, it will say Trying to backward through the graph a second time. Do you have any idea? Thanks!

**Comment by gpleiss** [comment]
@XiaoranHao can you open up a separate issue, and please include a code example?

**Comment by gpleiss** [comment]
@joseduc10 - I don't know if you still have this issue, but you would want to both use eval mode and the context:

```python
for i in range(niter):
    # Hadamard Multitask GP from 
    #https://docs.gpytorch.ai/en/latest/examples/03_Multitask_Exact_GPs/Hadamard_Multitask_GP_Regression.html
    model_gp.train()
    optimizer.zero_grad()
    outputs = model(full_train_x, full_train_i)
    loss_gp = -mll(output, y))
    # now sample from this model at regular intervals
    model.eval()
    with gpytorch.settings.detach_test_caches(False):
        output = model(regular_x, regular_i) 
        # run linear regression
        #TODO input to this model should actually be a set of samples from the GP
        y_pred = model_lr(outputs.mean)
        loss_lr = mse_loss(y_pred, y_true)
        
    # optimizer steps 
```

Closing this issue for now.
